system/lib/libc/dynlink.c

```diff
@@ -88,24 +96,24 @@ static struct dso* load_library_start(const char* name, int flags) {
   return p;
 }
 
-static void dlopen_js_onsuccess(void* handle) {
-  struct dso* p = (struct dso*)handle;
+static void dlopen_js_onsuccess(struct dso* dso, struct async_data* data) {
 #ifdef DYLINK_DEBUG
   fprintf(stderr, "%p: dlopen_js_onsuccess: dso=%p mem_addr=%p mem_size=%p\n", pthread_self(), p, p->mem_addr, p->mem_size);
 #endif
-  load_library_done(p);
+  load_library_done(dso);
   pthread_rwlock_unlock(&lock);
-  p->onsuccess(p->user_data, p);
+  data->onsuccess(data->user_data, dso);
+  free(data);
 }
 
-static void dlopen_js_onerror(void* handle) {
-  struct dso* p = (struct dso*)handle;
+static void dlopen_js_onerror(struct dso* dso, struct async_data* data) {
 #ifdef DYLINK_DEBUG
   fprintf(stderr, "%p: dlopen_js_onerror: dso=%p\n", pthread_self(), handle);
 #endif
   pthread_rwlock_unlock(&lock);
-  p->onerror(p->user_data);
-  free(p);
+  data->onerror(data->user_data);
```

Comment:
Doesn't onerror take 2 params like onsuccess? In JS they appear to.

Reply:
For theese internal JS callbacks they both take two arguments yes.    For those we use locally/internal `dlopen_js_onsuccess` and `dlopen_js_onerror`.

For the actual user callback passed to emscripten_dlopen (which is the user-facing ASI which doesn't change in this PR which is NFC) the onerror only takes one argument.   This is because when `emscripten_dlopen` is called there is no handle to a dll, and no handle is ever passed back to the user, except in the success case.

Reply:
I see now, thanks!